File: server/src/controller/user.controller.ts
import { Controller, Post, Put, Get, Body, Inject } from "@midwayjs/core";
import { Context } from "@midwayjs/koa";
import { ApiTags, ApiOperation } from "@midwayjs/swagger";
import { AuthService } from "../service/auth.service";
import { UserService } from "../service/user.service";
import { ResponseUtil } from "../common/types/response";
import { IUserContext } from "../interface";
import { Public } from "../decorator/public.decorator";
import {
  RegisterDTO,
  LoginDTO,
  UpdateProfileDTO,
  ChangePasswordDTO,
} from "../dto/user.dto";

@ApiTags("user")
@Controller("/api/users")
export class UserController {
  @Inject()
  authService!: AuthService;

  @Inject()
  userService!: UserService;

  @Public()
  @Post("/register")
  @ApiOperation({ summary: "用户注册" })
  async register(@Body() body: RegisterDTO) {
    const result = await this.authService.register(body);
    return ResponseUtil.success(result, "注册成功");
  }

  @Public()
  @Post("/login")
  @ApiOperation({ summary: "密码登录" })
  async login(@Body() body: LoginDTO) {
    const result = await this.authService.loginByPassword(
      body.phone,
      body.password,
    );
    return ResponseUtil.success(result, "登录成功");
  }

  @Get("/profile")
  @ApiOperation({ summary: "获取个人资料" })
  async getProfile(ctx: Context) {
    const user = ctx.state.user as IUserContext;
    const profile = await this.userService.findById(user.userId);
    if (!profile) {
      return ResponseUtil.error("用户不存在", 404);
    }
    const { password, ...safeProfile } = profile;
    return ResponseUtil.success(safeProfile);
  }

  @Put("/profile")
  @ApiOperation({ summary: "更新个人资料" })
  async updateProfile(ctx: Context, @Body() body: UpdateProfileDTO) {
    const user = ctx.state.user as IUserContext;
    const updated = await this.userService.updateProfile(user.userId, body);
    return ResponseUtil.success(updated, "更新成功");
  }

  @Post("/change-password")
  @ApiOperation({ summary: "修改密码" })
  async changePassword(ctx: Context, @Body() body: ChangePasswordDTO) {
    const user = ctx.state.user as IUserContext;
    await this.userService.changePassword(
      user.userId,
      body.oldPassword,
      body.newPassword,
    );
    return ResponseUtil.success(null, "密码修改成功");
  }

  @Public()
  @Post("/send-sms")
  @ApiOperation({ summary: "发送验证码" })
  async sendSms(@Body() body: { phone: string }) {
    const result = await this.authService.sendSmsCode(body.phone);
    return ResponseUtil.success(result, "验证码已发送");
  }

  @Public()
  @Post("/login-sms")
  @ApiOperation({ summary: "验证码登录" })
  async loginBySms(@Body() body: { phone: string; code: string }) {
    const result = await this.authService.loginBySms(body.phone, body.code);
    return ResponseUtil.success(result, "登录成功");
  }
}

File: server-java/src/main/java/com/wudong/controller/UserController.java
package com.wudong.controller;

import com.wudong.common.response.ApiResponse;
import com.wudong.dto.user.*;
import com.wudong.entity.User;
import com.wudong.security.SecurityUtils;
import com.wudong.service.AuthService;
import com.wudong.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@Tag(name = "用户")
@RestController
@RequestMapping("/api/users")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;
    private final AuthService authService;

    @Operation(summary = "用户注册")
    @PostMapping("/register")
    public ApiResponse<Map<String, Object>> register(@Valid @RequestBody RegisterRequest request) {
        User user = userService.register(request.getPhone(), request.getPassword(), request.getNickname());
        Map<String, Object> result = userService.loginByPassword(request.getPhone(), request.getPassword());
        return ApiResponse.success(result, "注册成功");
    }

    @Operation(summary = "密码登录")
    @PostMapping("/login")
    public ApiResponse<Map<String, Object>> login(@Valid @RequestBody LoginRequest request) {
        Map<String, Object> result = userService.loginByPassword(request.getPhone(), request.getPassword());
        return ApiResponse.success(result, "登录成功");
    }

    @Operation(summary = "发送短信验证码")
    @PostMapping("/send-sms")
    public ApiResponse<Void> sendSms(@Valid @RequestBody SendSmsRequest request) {
        authService.sendSms(request.getPhone());
        return ApiResponse.success(null, "验证码已发送");
    }

    @Operation(summary = "短信验证码登录")
    @PostMapping("/login-sms")
    public ApiResponse<Map<String, Object>> loginBySms(@Valid @RequestBody LoginSmsRequest request) {
        authService.verifySms(request.getPhone(), request.getCode());
        Map<String, Object> result = userService.loginBySms(request.getPhone());
        return ApiResponse.success(result, "登录成功");
    }

    @Operation(summary = "获取用户信息")
    @GetMapping("/profile")
    public ApiResponse<User> getProfile() {
        Long userId = SecurityUtils.getCurrentUserId();
        User user = userService.getUserById(userId);
        return ApiResponse.success(user);
    }

    @Operation(summary = "更新用户信息")
    @PutMapping("/profile")
    public ApiResponse<User> updateProfile(@Valid @RequestBody UpdateProfileRequest request) {
        Long userId = SecurityUtils.getCurrentUserId();
        User user = userService.updateProfile(userId, request.getNickname(), request.getAvatar(),
                request.getGender(), request.getRegion(), request.getBio());
        return ApiResponse.success(user, "更新成功");
    }

    @Operation(summary = "修改密码")
    @PostMapping("/change-password")
    public ApiResponse<Void> changePassword(@Valid @RequestBody ChangePasswordRequest request) {
        Long userId = SecurityUtils.getCurrentUserId();
        userService.changePassword(userId, request.getOldPassword(), request.getNewPassword());
        return ApiResponse.success(null, "密码修改成功");
    }
}
